app/proxy/handlers.go

```diff
@@ -65,49 +65,92 @@ func accessLogHandler(wr io.Writer) func(next http.Handler) http.Handler {
 
 func stdoutLogHandler(enable bool, lh func(next http.Handler) http.Handler) func(next http.Handler) http.Handler {
 
-	if enable {
-		log.Printf("[DEBUG] stdout logging enabled")
-		return func(next http.Handler) http.Handler {
-			fn := func(w http.ResponseWriter, r *http.Request) {
-				// don't log to stdout GET ~/(.*)/ping$ requests
-				if r.Method == "GET" && strings.HasSuffix(r.URL.Path, "/ping") {
-					next.ServeHTTP(w, r)
-					return
-				}
-				lh(next).ServeHTTP(w, r)
-			}
-			return http.HandlerFunc(fn)
-		}
+	if !enable {
+		return passThroughHandler
 	}
 
+	log.Printf("[DEBUG] stdout logging enabled")
 	return func(next http.Handler) http.Handler {
-		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-			next.ServeHTTP(w, r)
-		})
+		fn := func(w http.ResponseWriter, r *http.Request) {
+			// don't log to stdout GET ~/(.*)/ping$ requests
+			if r.Method == "GET" && strings.HasSuffix(r.URL.Path, "/ping") {
+				next.ServeHTTP(w, r)
+				return
+			}
+			lh(next).ServeHTTP(w, r)
+		}
+		return http.HandlerFunc(fn)
 	}
 }
 
 func gzipHandler(enabled bool) func(next http.Handler) http.Handler {
-	if enabled {
-		log.Printf("[DEBUG] gzip enabled")
-		return handlers.CompressHandler
+	if !enabled {
+		return passThroughHandler
 	}
 
-	return func(next http.Handler) http.Handler {
-		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-			next.ServeHTTP(w, r)
-		})
-	}
+	log.Printf("[DEBUG] gzip enabled")
+	return handlers.CompressHandler
 }
 
 func signatureHandler(enabled bool, version string) func(next http.Handler) http.Handler {
-	if enabled {
-		log.Printf("[DEBUG] signature headers enabled")
-		return R.AppInfo("reproxy", "umputun", version)
+	if !enabled {
+		return passThroughHandler
 	}
-	return func(next http.Handler) http.Handler {
-		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-			next.ServeHTTP(w, r)
-		})
+	log.Printf("[DEBUG] signature headers enabled")
+	return R.AppInfo("reproxy", "umputun", version)
+}
+
+// limiterSystemHandler throttles overall activity of reproxy server, 0 means disabled
+func limiterSystemHandler(reqSec int) func(next http.Handler) http.Handler {
+	if reqSec <= 0 {
+		return passThroughHandler
+	}
+	return func(h http.Handler) http.Handler {
+		lmt := tollbooth.NewLimiter(float64(reqSec), nil)
+		fn := func(w http.ResponseWriter, r *http.Request) {
+			if httpError := tollbooth.LimitByKeys(lmt, []string{"system"}); httpError != nil {
+				http.Error(w, http.StatusText(http.StatusTooManyRequests), http.StatusTooManyRequests)
```

Comment:
I see what you mean in both comments, good points. But I think we may ignore both issues for now and undercount hits in those cases. Practically it will limit excessive activity good enough, not as precise as it can but still ok.